### PR TITLE
Improve op_helper - Clean helpers

### PR DESCRIPTION
**Pull request by latot**
Fixes https://github.com/cbm755/octsympy/issues/639

Hi, well, now this is a little clean in the helpers, and a generalization of it, so we will no need rewrite new functions for this, this code practically don't affect performance, with relative low interaction (less than 1000 operations) are very similar, some times one is faster other moments the other, while with a 10000 operations we finally can see we have an approximation of 6% of less performance.

Other point is with this is defined the operation element by element to any function, and if we found some problem we only need fix one file instead of 3.

I write this code in this way thinking in Tensor.

Thx. Cya.

### CODE REVIEW

**Comment by latot** [comment]
I change the round function, because in this issue https://github.com/cbm755/octsympy/issues/635, we found two problems, and this change expose one of it, so its necessary to pass the tests, and this fix don't solve the the issue completely.

File: inst/@sym/private/op_helper.m [hunk 1]
```diff
@@ -44,19 +43,20 @@
...
-          '    return _op(x, y),' ];
+          'l = (1, 1)'
+          'for i in range(len(_ins)):'
+          '    if isinstance(_ins[i], MatrixBase):'
+          '        if _ins[i].shape != (1, 1):'
+          '            l = _ins[i].shape'
+          '        else:'
+          '            assert l == _ins[i].shape, "Matrix must have equal sizes"'
+          '    else:'
+          '        _ins[i] = Matrix([_ins[i]])'
```
Comment (cbm755) [inline inst/@sym/private/op_helper.m]: I think you should first convert it to matrix, then check the size is either same or (1,1).

File: inst/@sym/private/op_helper.m [hunk 2]
```diff
@@ -44,19 +43,20 @@
...
+          'base = zeros(*l)'
+          'for i in itertools.product(*map(Range, l)):'
+          '    base.__setitem__(i, _op(*[k[0] if k.shape == (1, 1) else k.__getitem__(i) for k in _ins]))'
+          'return base,' ];
 
-  z = python_cmd (cmd, sym(x), sym(y));
+  z = python_cmd (cmd, sym(varargin){:});
```
Comment (cbm755) [inline inst/@sym/private/op_helper.m]: Can't change indexing on Matlab (I think).  Also, IIRC, we might deprecate `sym(cellarray)`.  Can you use `cellfun` or something to do it?

File: inst/@sym/private/op_helper.m [hunk 3]
```diff
@@ -44,19 +43,20 @@
...
+          'for i in range(len(_ins)):'
+          '    if isinstance(_ins[i], MatrixBase):'
+          '        if _ins[i].shape != (1, 1):'
+          '            l = _ins[i].shape'
+          '        else:'
+          '            assert l == _ins[i].shape, "Matrix must have equal sizes"'
+          '    else:'
+          '        _ins[i] = Matrix([_ins[i]])'
+          'base = zeros(*l)'
+          'for i in itertools.product(*map(Range, l)):'
```
Comment (cbm755) [inline inst/@sym/private/op_helper.m]: Can this just be:
````
for i in range(0, len(<something>)):
    base[i] = ...
````

File: inst/@sym/abs.m [hunk 4]
```diff
@@ -40,7 +40,7 @@
   if (nargin ~= 1)
     print_usage ();
   end
-  y = uniop_helper (x, 'Abs');
+  y = op_helper ('Abs', x);
```
Comment (cbm755) [inline inst/@sym/abs.m]: For many of these files, the change needs to be made in `generate_functions.py` (see note in the help text).

File: inst/@sym/private/op_helper.m (hunk 2 above)
Comment (cbm755) [inline inst/@sym/private/op_helper.m]: Sorry I meant "can't chain indexing in Matlab".

File: inst/@sym/private/op_helper.m (hunk 3 above)
Comment (latot) [inline inst/@sym/private/op_helper.m]: Hi, the send line works, while the first sadly not, that only will travel in a 1 dim matrix.
The idea of the line is travel in all values of the matrix, while the matrix can have any number of dims.

**Comment by cbm755** [comment]
> only will travel in a 1 dim matrix

Seems to work for me:
````
In [21]: A = Matrix([[x,2,3],[3,4,5]])

In [22]: B = zeros(*A.shape)

In [23]: for i in range(0, len(A)):
   ....:     B[i] = 2*A[i]
   ....:     

In [24]: B
Out[24]: 
Matrix([
[2*x, 4,  6],
[  6, 8, 10]])
````

File: inst/@sym/private/op_helper.m (hunk 2 above)
Comment (latot) [inline inst/@sym/private/op_helper.m]: Hi, i split the chain, about cellfun sadly i can't use it, at least in a right way, because it converts the cells to matrix! and with it we can't unpack it to a function. Obvs i could researched bad, if you know some way please tell me and i'll update this.
In some way i prefer do this in other pr, because i think we should update all the code with that instead only a few lines.

If don't exist, maybe use a helper instead?

Thx. Cya.

File: inst/@sym/eq.m [hunk 5]
```diff
@@ -88,7 +88,7 @@
     print_usage ();
   end
 
-  t = ineq_helper('[donotuse]', 'Eq', x, y);
+  t = ineq_helper('[donotuse]', 'Eq', sym(x), sym(y));
```
Comment (cbm755) [inline inst/@sym/eq.m]: this change is unrelated to this PR?

File: inst/@sym/rewrite.m [hunk 6]
```diff
@@ -109,7 +109,7 @@
 
   F = python_cmd(cmd, sym(f), how);
 
-  % maintainer note: binop_helper would sym(how)
+  % maintainer note: op_helper would sym(how)
```
Comment (cbm755) [inline inst/@sym/rewrite.m]: ... for example, this is not true any more: maybe just say "op_helper might sym(how)"

**Comment by latot** [comment]
You are right, sorry i misread what you mean, now i change it.

With last change now the helper support more formats, not only sym objects, its useful to can simplify other functions like sym, so now we don't transform the objects to sym in the helper and we need declare it from the functions, this is why i update the functions with ineq_helper (ineq_helper call op_helper).

Thx. Cya.

File: inst/@sym/isempty.m [hunk 7]
```diff
@@ -60,7 +60,6 @@
 %! a = sym ([1 2]);
 %!assert (~isempty (sym (1)))
 %!assert (isempty (sym (se)))
-%!assert (isempty (se == []))
```
Comment (cbm755) [inline inst/@sym/isempty.m]: WTF?

File: inst/@sym/laguerreL.m [hunk 8]
```diff
@@ -63,10 +63,9 @@
 function L = laguerreL(n, alpha, x)
 
   if (nargin == 2)
-    x = alpha;
```
Comment (cbm755) [inline inst/@sym/laguerreL.m]: I want this back, too hard to read the code otherwise.

File: inst/@sym/lambertw.m [hunk 9]
```diff
@@ -52,17 +52,12 @@
...
-    x = sym(x);
-    k = sym(k);
-    W = binop_helper (x, k, 'LambertW');
-  else
+function W = lambertw(varargin)
+  if (nargin > 2)
     print_usage ();
   end
+  varargin = sym(varargin);
+  W = op_helper ('lambda a: LambertW(*reversed(a))', varargin);
```
Comment (cbm755) [inline inst/@sym/lambertw.m]: nice!

File: inst/@sym/limit.m [hunk 10]
```diff
@@ -83,14 +83,7 @@
...
 
-  cmd = { '(f, x, a, pdir) = _ins'
-          '# note, not MatrixExpr'
-          'if isinstance(f, sp.MatrixBase):'
-          '    g = f.applyfunc(lambda b: b.limit(x, a, dir=pdir))'
-          'else:'
-          '    g = f.limit(x, a, dir=pdir)'
-          'return g,' };
-  L = python_cmd (cmd, sym(f), sym(x), sym(a), pdir);
+  L = op_helper('lambda a, b, c, d: a.limit(b, c, d)', sym(f), sym(x), sym(a), pdir);
```
Comment (cbm755) [inline inst/@sym/limit.m]: this must be another PR

**Comment by cbm755** [comment]
I started reviewed agian, but there is too much new stuff here!!  This was a nice simple PR that I as looking forward to merging.  Please  keep it that way.  

EACH PR DOES ONE SIMPLE THING!  That way, it will get reviewed and merged.  This one will sit here until it looks simple again.

**Comment by latot** [comment]
Okidokis!, splitting....

File: inst/@sym/isempty.m (hunk 7 above)
Comment (latot) [inline inst/@sym/isempty.m]: that don't have any sense..., because [] is equal to []..., why it should return empty? (se is [])

**Comment by latot** [comment]
Hi!, i split this, when this will be merged i'll send the next pr with the changes.
I add documentation to the helper.

Thx. Cya.

File: inst/@sym/isempty.m (hunk 7 above)
Comment (cbm755) [inline inst/@sym/isempty.m]: elementwise comparision I suppose.  But even if the test is wrong, its not this PRs position to change it.  If the test is wrong, do another PR.

Unless this PR changes behaviour in some way that the test used to pass and now it fails...  But if so I'm worried because this PR is just a code refactor.

File: inst/@sym/private/op_helper.m [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+    %assert strncmp(scalar_fcn_str, 'def ', 4)
+    cmd = scalar_fcn;
+  else
+    cmd = {['_op = ' scalar_fcn]};
+  end
+
+  % note: cmd is already cell array, hence [ concatenates with it
+  cmd = [ cmd
+          'q = Matrix([0])'
+          'for i in _ins:'
```
Comment (cbm755) [inline inst/@sym/private/op_helper.m]: maybe `for A in _ins:` better  (`i` is for integer)

**Comment by cbm755** [comment]
Sorry for losing temper earlier: had only a few mins to review and was sad that it looked so different.

Now looks much better!  Any idea why lambertw is failing?

**Comment by latot** [comment]
Hi!, i restore lambertw function but i forgot update the function, now should works, i'm actually with a few mins with internet so i can't check if travis pass, my hope is it works.

Thx. Cya.

File: inst/@sym/private/op_helper.m (hunk 11 above)
Comment (latot) [inline inst/@sym/private/op_helper.m]: Okidokis

**Comment by cbm755** [comment]
Ok, this looks good.  Feel free to prepare those other fixes as new PRs but let's try to keep each self-contained.

File: inst/@sym/isempty.m (hunk 7 above)
Comment (cbm755) [inline inst/@sym/isempty.m]: Oops, I missed that this was still here before I merged :(

I will continue this discussion at #650
